torch/csrc/jit/codegen/cuda/lower2device.cpp

```diff
@@ -121,12 +209,135 @@ std::ostream& GPULower::printKernel(
   return os;
 }
 
-std::string GPULower::getKernel(const std::string& kernel_name) {
+std::string GpuLower::getKernel(const std::string& kernel_name) {
   std::stringstream ss;
   printKernel(ss, kernel_name);
   return ss.str();
 }
 
+// Maps Fusion IR nodes to the Kernel IR counterparts
+// (this is a interim solution for easing the Kernel IR splitting)
+class TORCH_CUDA_API GpuLower::KernelIrMapper : private OptInConstDispatch {
+ public:
+  explicit KernelIrMapper(GpuLower* gpu_lower) : gpu_lower_(gpu_lower) {}
+
+  Val* lower(const Val* value) {
+    const auto it = gpu_lower_->kir_map_.find(value);
+    if (it != gpu_lower_->kir_map_.end()) {
+      return it->second;
+    } else {
+      handle(value);
+      const auto lowered_node = gpu_lower_->kir_map_[value];
+      TORCH_CHECK(lowered_node != nullptr);
+      TORCH_CHECK(kir::isLoweredVal(lowered_node));
+
+      // Lower the arithmetic expression defining the value, if any
+      if (value->isScalar()) {
+        if (auto def = value->getOrigin()) {
+          lowerDefinition(lowered_node, def);
+        }
+      }
+
+      return lowered_node;
+    }
+  }
+
+ private:
+  // TODO(kir): rewrite this
+  void lowerDefinition(Val* lowered_value, const Expr* def) {
+    switch (def->type()) {
+      case ExprType::UnaryOp: {
+        const auto op = def->as<fuser::UnaryOp>();
+        new kir::UnaryOp(op->getUnaryOpType(), lowered_value, lower(op->in()));
+        break;
+      }
+      case ExprType::BinaryOp: {
+        const auto op = def->as<fuser::BinaryOp>();
+        new kir::BinaryOp(
+            op->getBinaryOpType(),
+            lowered_value,
+            lower(op->lhs()),
+            lower(op->rhs()));
+        break;
+      }
+      case ExprType::TernaryOp: {
+        const auto op = def->as<fuser::TernaryOp>();
+        new kir::TernaryOp(
+            op->getTernaryOpType(),
+            lowered_value,
+            lower(op->in1()),
+            lower(op->in2()),
+            lower(op->in3()));
+        break;
+      }
+      default:
+        TORCH_CHECK(false, "Unexpected expression type");
+    }
+  }
+
+  void handle(const Statement* node) override {
+    OptInConstDispatch::handle(node);
+  }
+
+  void handle(const Val* node) override {
+    OptInConstDispatch::handle(node);
+  }
+
+  void handle(const Expr* node) override {
+    OptInConstDispatch::handle(node);
+  }
+
+  void handle(const TensorDomain* node) override {
+    const auto lowered_node = new kir::TensorDomain(node);
+    TORCH_CHECK(gpu_lower_->kir_map_.insert({node, lowered_node}).second);
```

Comment:
Why not put `gpu_lower_->kir_map_.insert({node, lowered_node}).second` in the constructor `new kir::TensorDomain(node)`

I'm worried people could call `kir::TensorDomain(node)` directly instead of using `loweredValue(node)` which would typically be a bad thing to do.

Reply:
`kir::TensorDomain`'s constructor doesn't know about GpuLower (or have access to kir_map_). Or have I misunderstood the comment?

I do agree about the potential trouble with people calling the "lowering" constructors directly though. I'm still considering options to limit this, but I haven't spent too much time on it since we need to revisit the whole Kernel IR "builder" interface soon anyway (once it's a separate class hierarchy, we're not going to have the luxury of FusionGuard or Fusion::registerStuff())

Reply:
Sounds good, I'm hopeful we can limit it, I got in some trouble when they were called explicitly before, would be nice to hide them if we can.